src/go/k8s/pkg/resources/statefulset_update.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2021 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+// Package resources contains reconciliation logic for redpanda.vectorized.io CRD
+package resources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const requeueDuration = time.Second * 10
+
+func (r *StatefulSetResource) updateStsImage(
```

Comment:
I see this function differently. To embrace returning early and don't nesting unnecessarily the pseudo code should look like:
``` go
func (r *StatefulSetResource) updateStsImage(ctx context.Context, sts *appsv1.StatefulSet,) error {
    if images in sts and cluster CR are the same and ready replicas are the same as cluster.spec.replicas {
        return
    }
    
    set the upgrade status in the cluster CR
    
    modifyPodImage
    
    lunchOnePodUpdate using `StatefulSet's partition number`

    return to reschedule
}
```

We can argue that setting upgrade status and modify pod image could be done once. Never the less this should be the core logic behind this function.

Reply:
The replicas condition wouldn't work. I explained in an earlier comment-reply you made. Beyond the if-statement, what you describe is the logic currently implemented.